### DISCUSSION TITLE
State Transition Algorithm

### DISCUSSION THREAD

**Discussion by meganindya** [sugarlabs/musicblocks-v4#7] [Ideas]
A _state_ is basically an exhaustive set of all fundamental properties, e.g. ADSR properties (pitch, sustain, synth, etc.), pen properties (pen color, thickness, etc.), etc. There is a _global state_ which are the defaults, and each "mouse" gets a copy of the _global state_ which it overrides as it executes instructions.

Every _clamp_ block saves a copy of the _contextual state_ (_transitory state_) before it, so that its nested instructions can override that, and the "mouse" can return back to the _contextual state_ after the nested blocks have been executed. An instruction which executes a _callback_ (_action_) does the same. There need not be special cases for _start_ and _action_ blocks as they are _clamp_ blocks too. The _start_ blocks just begins with the _global state_ and the _action_ blocks receive its context from the caller.

## State Transition Diagram (NFA)

![State Transition Diagram](https://user-images.githubusercontent.com/33328728/102010645-69be4300-3d65-11eb-8bd3-1513772aeab8.png)

[View Lucidchart Diagram](https://lucid.app/documents/view/915afcd8-9ee0-42d9-a42e-9a20073fc781)

@walterbender this should be way faster than evaluating each property every single time there is a _note_ block.

**Comment by walterbender** [comment]
and it will make all of the logic around clamp blocks more consistent and easier to maintain. Very nice.

**Reply by meganindya** [reply]
no more hidden blocks :D

**Comment by meganindya** [comment]
We needed a meeting. I sent a thread on slack. Sometime later today would be possible?

**Comment by walterbender** [comment]
in a couple of hours? (My 10:30 AM ?)

**Comment by meganindya** [comment]
works for me.
